client-lite/src/trace/do_log.cpp

```diff
@@ -0,0 +1,437 @@
+            case Level::Warning:    return "warn ";
+            case Level::Info:       return "info ";
+            case Level::Verbose:    return "     ";
+        }
+        return "";
+    }
+
+    static int s_LogFilesFilter(const struct dirent* logfile)
+    {
+        return (logfile->d_type == DT_REG) && (strstr(logfile->d_name, g_logFileNamePrefix) != NULL);
+    }
+
+    void _ClearOlderLogFilesIfNeeded()
+    {
+        // From: https://github.com/Azure/iot-hub-device-update/blob/main/src/logging/zlog/src/zlog.c
+
+        // List the files specified by file_select in alphabetical order
+        struct dirent** logfiles;
+        const int nFiles = scandir(_logDir.c_str(), &logfiles, s_LogFilesFilter, alphasort);
+        if (nFiles == -1)
+        {
+            return;
+        }
+
+        if (static_cast<uint32_t>(nFiles) > g_maxLogFiles)
+        {
+            for (uint32_t i = 0; i < (static_cast<uint32_t>(nFiles) - g_maxLogFiles); ++i)
+            {
+                char filepath[512];
+                if (SUCCEEDED(StringPrintf(filepath, ARRAYSIZE(filepath), "%s/%s", _logDir.c_str(), logfiles[i]->d_name)))
+                {
+                    remove(filepath);
+                    ++_stats.numFilesDeleted;
+                }
+            }
+        }
+
+        // Free memory allocated by scandir.
+        for (int i = 0; i < nFiles; ++i)
+        {
+            free(logfiles[i]);
+        }
+        free(logfiles);
+    }
+
+    void _CreateLogFile()
+    {
+        ++_stats.numFileCreationAttempts;
+
+        const auto timestamp = SysTimePointToFileNameString(wall_clock_t::now());
+        char filePath[512];
+        const auto hr = StringPrintf(filePath, ARRAYSIZE(filePath), "%s/%s%s.log", _logDir.c_str(), g_logFileNamePrefix, timestamp.data());
+        DO_ASSERT(SUCCEEDED(hr));
+
+        _logFile.open(filePath, std::fstream::out | std::fstream::app);
+        if (_logFile.fail())
+        {
+            fprintf(stderr, "Failed to create log file at %s", filePath);
+        }
+        else
+        {
+            ++_stats.numFilesCreated;
+        }
+    }
+
+    void _RotateLogFilesIfNeeded()
+    {
+        const bool fLogWasOpen = _logFile.is_open();
+        if (fLogWasOpen && (_logFile.tellp() >= static_cast<std::streampos>(g_maxLogFileSizeBytes)))
+        {
+            _logFile.close();
+        }
+        _ClearOlderLogFilesIfNeeded();
+        if (fLogWasOpen && !_logFile.is_open())
+        {
+            _CreateLogFile();
+        }
+    }
+
+    void _FlushThreadProc()
+    {
+        while (true)
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            _cv.wait_for(lock, std::chrono::minutes(3), [this](){ return !_fRunning; });
+            if (!_fRunning)
+            {
+                break;
+            }
+
+            _writeBuffer.Flush(_logFile);
+            _RotateLogFilesIfNeeded();
+            ++_stats.numFlushThreadLoops;
```

Comment:
Would it also be beneficial to have a max memory limit for logging (incase of agent business -> logging a ton of stuff?) i.e. flush if we're using x memory

Reply:
Yes this is done already. The LogBuffer is a fixed size buffer (64KB right now). LoggerImpl::WriteV method performs the flush in case buffer gets full within this 3 min interval.